### ISSUE TITLE
docker-compose.override.yml and Expanding the Station Port Range

### ISSUE THREAD

**Issue by JohnyBeGood** [AzuraCast/AzuraCast#5385]
### Installation Method

One-Click Installation (Docker)

### AzuraCast Release Channel

Stable Channel

### Current AzuraCast Version

v0.16.1 Stable

### What happened?

Hey all,

I'm on Debian 10 and in order for me to move away from CentovaCast I need to have ports starting from 7500.
So, I followed [Expanding the Station Port Range](https://docs.azuracast.com/en/administration/docker) and I created **docker-compose.override.yml** using vim just to make sure there was no formatting issues. I also changed **azuracast.env** **AUTO_ASSIGN_PORT_MIN=7500** but as soon as I try to restart I get this error
```
root@azuracast:/var/azuracast# docker-compose down
service "stations" has neither an image nor a build context specified: invalid compose project
root@azuracast:/var/azuracast#
```
It seems as soon as create the file it starts to have problem.

I don't see option to specify station streaming ports here as suggested:

```
root@azuracast:/var/azuracast# ./docker.sh update-self
New Docker utility script downloaded.
root@azuracast:/var/azuracast# ./docker.sh change-ports
Port to use for HTTP connections [80]: 
Port to use for HTTPS connections [443]: 
Port to use for SFTP connections [2022]: 
[+] Running 2/2
 ⠿ Container azuracast        Removed                                                                                                                                               20.8s
 ⠿ Network azuracast_default  Removed                                                                                                                                                0.2s
[+] Running 2/2
 ⠿ Network azuracast_default  Created                                                                                                                                                0.1s
 ⠿ Container azuracast        Started                                                                                                                                                3.8s
root@azuracast:/var/azuracast# 
```

Any suggestions?

**Comment by Vaalyn** [comment]
Hey 👋 

we are currently in the process of updating our docs to the new format for such configurations since we merged our multiple containers into a single one.

Here is what your `docker-compose.override.yml` should look like now:
```yml
services:
  web:
    ports:
      - "8500-8999:8500-8999"
```

@SC2Mitch just to be sure, do you have this part of the docs already on your list of changes needed for the docs?

**Comment by SC2Mitch** [comment]
> @SC2Mitch just to be sure, do you have this part of the docs already on your list of changes needed for the docs?

I believe it's in my 0.16 branch but I'll check tonight, if not I'll add it in

**Comment by SC2Mitch** [comment]
Verify that nothing else is running on the ports you're attempting to use.
Otherwise you can restart docker and reattempt to do what you was doing.

**Comment by celilo** [comment]
>  Did you get the same error when starting the container? `Error response from daemon: driver failed programming external connectivity on endpoint azuracast`

No, but I included the access ports in my docker-compose.override.yml. 
For simplicity, try putting this in your docker-compose.override.yml  
```
 ports:
       - '${AZURACAST_HTTP_PORT:-80}:80'
       - '${AZURACAST_HTTPS_PORT:-443}:443'
       - '${AZURACAST_SFTP_PORT:-2022}:2022'
       - "7500-7999:7500-7999"
       # This allows Azuracast to use the above series of ports, which should not conflict with the 8000+ range used by 
       # CentovaCast. I don't know if you are running anything else that would conflict with that range.

```
Also change the the min and max ports in azuracast.env
```
AUTO_ASSIGN_PORT_MIN=7500
AUTO_ASSIGN_PORT_MAX=7999
```
Then
```
docker-compose down
docker-compose up -d
```

**Comment by JohnyBeGood** [comment]
Thanks for the replies!

Even before trying [celilo](https://github.com/celilo) suggestion, I was getting this:
```
root@azuracast:~# docker-compose down
no configuration file provided: not found
root@azuracast:~# docker-compose up -d
no configuration file provided: not found
root@azuracast:~#
```
Did reboot, same error. Other than AzuraCast I do not have any other services running on this server, not even CentovaCast, this is a fresh install.

This is how my **docker-compose.override.yml** looks like:

```
services:
  web:
     ports:
       - '${AZURACAST_HTTP_PORT:-80}:80'
       - '${AZURACAST_HTTPS_PORT:-443}:443'
       - '${AZURACAST_SFTP_PORT:-2022}:2022'
       - "7500-7999:7500-7999"
       # This allows Azuracast to use the above series of ports, which should not conflict with the 8000+ range used by 
       # CentovaCast. I don't know if you are running anything else that would conflict with that range.

```
and MIN/MIN port section:


```
# Minimum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8000
AUTO_ASSIGN_PORT_MIN=7500

# Maximum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8499
AUTO_ASSIGN_PORT_MAX=7999
```

**Comment by celilo** [comment]
```
root@azuracast:~# docker-compose down
no configuration file provided: not found
root@azuracast:~# docker-compose up -d
no configuration file provided: not found
root@azuracast:~#
```
Looks like you are in the wrong directory. If you followed the install instructions:
`cd /var/azuracast`

**Comment by celilo** [comment]
Also make sure that your modified files are int he same folder.

**Comment by JohnyBeGood** [comment]
Grrr, I guess when I was making reboots I managed to forget to switch to the directory :(
Now even when I'm in that directory I get the same error:

```
root@azuracast:/var/azuracast# docker-compose down
[+] Running 2/2
 ⠿ Container azuracast        Removed                                                                                                                                                0.2s
 ⠿ Network azuracast_default  Removed                                                                                                                                                0.2s
root@azuracast:/var/azuracast# docker-compose up -d
[+] Running 1/2
 ⠿ Network azuracast_default  Created                                                                                                                                                0.1s
 ⠿ Container azuracast        Starting                                                                                                                                             130.1s
Error response from daemon: driver failed programming external connectivity on endpoint azuracast (b5a364f7916e8ce7538c23bc466ff5d03c14897ff723339553874c08d1168413): Error starting userland proxy:
root@azuracast:/var/azuracast# 
```

**Comment by celilo** [comment]
My apologies. Most of what was written is accurate with one exception that will prevent the desired outcome. The override file is additive on merges.
https://docs.docker.com/compose/extends

I had already modified the offending file in my process, so I didn't initially catch this. I would still recommend use of the override file, but you will have to remove the offending ports from the main docker-compose.yml file. So you can still keep your custom ports in the override file so that you can simply delete the unneeded ports int the main docker-compose.yml file each time it is updated.

So use Vaalyn's format with your desired ports, like below:
```
services:
  web:
    ports:
      - "7500-7999:7500-7999"
```
In summary, If you leave the access ports in the main docker-compose file, they will be merged, so no need to add them in override. So, if you ant to add ports, the override file alone is great, but you need to remove the ports from the main file if you do not want to use them.

**Comment by JohnyBeGood** [comment]
I think I got it to work! 
The odd part is that I'm no longer getting Those errors "Error response from daemon: driver failed programming external..." Only this time I would stop the container with docker-compose down and start it again once I was done editing files. So far, I only added this range to **docker-compose.override.yml**
```
services:
  web:
    ports:
      - "7500-7550:7500-7550"
```
and also enabled below in **azuracast.env** since I need a small range of ports:

```
# Minimum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8000
AUTO_ASSIGN_PORT_MIN=7500

# Maximum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8499
AUTO_ASSIGN_PORT_MAX=7550
```
Did reboot a few times just to make sure and started the container. Is this only needed to survive future container updates?

**Comment by celilo** [comment]
The override file will survive the container updates.
You originally indicated that you didn't want to use ports in the low 8000 range, but you left those ports int he docker-compose file. I had to remove those ports from the docker compose file to prevent a conflict with another docker container running on port 8000. Edits, such as mine, to the docker-compose file will not survive a container update.
If you don't have any other containers running on those ports,, then you don't need to worry about it. This may be the case if you simply wanted to avoid using duplicate ports for your stations served by CentovaCast from a different device.

**Comment by JohnyBeGood** [comment]
Thanks for the reply!

Sorry for not being clear. CentovaCast on other server has 7500-7550 range that I'm trying to mimic and when I test it out I'm trying to move over station to AzuraCast.
Initially ie. 7520 port was not allowed on AzuraCast. 
This is only container on this machine. No other services are running on this server so that's a good news!

**Comment by JohnyBeGood** [comment]
I guess I spoke too soon :(
Shoutcast admin does not work with adding additional ports. I even removed everything and followed documentation and only left the following in **docker-compose.override.yml**:
```
services:
  web:
    ports:
      - "8500-8999:8500-8999"
```
and in **azuracast.env** these ranges:

```
# Minimum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8000
AUTO_ASSIGN_PORT_MIN=8500

# Maximum Port for Station Port Assignment
# Modify this if your stations are listening on nonstandard ports.
# Default: 8499
AUTO_ASSIGN_PORT_MAX=8999
```

Accessing the admin interface on port 8040 works, but on 8504 does not work?

![AC_station_page_SC_admin_port_8040_works](https://user-images.githubusercontent.com/1755800/170579788-4d96e781-a247-450f-9a09-d8edfd14dad4.jpg)
![shoutcast_logged_in](https://user-images.githubusercontent.com/1755800/170579800-8442cb3f-b9da-4e51-88d9-ee2ee54511d8.jpg)
![AC_station_page_SC_admin_port_8504_link_does_not_work](https://user-images.githubusercontent.com/1755800/170579797-eb510429-e403-40cd-b759-f50334a35df9.jpg)
![shoutcast_unable_to_open_admin_gui](https://user-images.githubusercontent.com/1755800/170579803-3df7f729-b2ea-40f7-a853-dc8366dd4464.jpg)

**Comment by celilo** [comment]
I was running an Icecast back-end, so I can't speak to your Shoutcast issue. I've also moved on from Azuracast because I need real-time support, but hopefully somebody else can help you. I've removed my installation, so I can't easily test.

**Comment by JohnyBeGood** [comment]
> I was running an Icecast back-end, so I can't speak to your Shoutcast issue. I've also moved on from Azuracast because I need real-time support, but hopefully somebody else can help you. I've removed my installation, so I can't easily test.

I see. No worries. Hopefully devs will chime in.
Thanks for taking the time to reply!

**Comment by JohnyBeGood** [comment]
Any updates on this before it automatically closes?

**Comment by JohnyBeGood** [comment]
Anyone?
